### PR TITLE
RQM Changes for Dynamic Field Mapping

### PR DESCRIPTION
**Pull request by prashelke**
Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] PR description and commit message should describe the changes done in this PR
- [x] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [x] Latest Code from master or specific release branch is merged to your branch
- [x] No unwanted\commented\junk code is included
- [x] No new warning upon build solution
- [x] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [x] Verify that changes are compatible with all relevant browsers and platforms.
- [x] After creating pull request there should not be any conflicts
- [ ] Resolve all Codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Release Notes

- **New Features**
  - Integration of Rational Quality Manager (RQM) functionalities into the ALMIntegration class.
  - Enhanced export capabilities for business flows to RQM, including handling of external item fields.
  - New methods for monitoring network logs in the Selenium driver.

- **Bug Fixes**
  - Improved error handling and user feedback for ALM project connections and disconnections.

- **Enhancements**
  - Added properties for custom fields in various classes to better manage external item data.
  - Refined logic for exporting and importing item fields, enhancing integration with RQM.
  - Improved handling of external item fields and custom attributes in the RQM integration.

- **Documentation**
  - Updated method signatures and added comments for clarity on new functionalities.

### CODE REVIEW

File: Ginger/Ginger/ALM/Repository/RQMRepository.cs [hunk 1]
```diff
@@ -240,7 +241,35 @@
...
+                    {
+                        externalItemFieldBasetest.SelectedValue = externalItemFieldBase.SelectedValue;
+                    }
+                    OriginalExternalFieldBase.Add(externalItemFieldBasetest);
+                }
+            }
+            else
+            {
+                OriginalExternalFieldBase = WorkSpace.Instance.Solution.ExternalItemsFields;
+            }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/ALM/Repository/RQMRepository.cs]: Done

File: Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs [hunk 2]
```diff
@@ -1227,11 +1243,265 @@
...
+
+                            // Set the first item as SelectedValue if PossibleValues is not empty
+                            if (field.PossibleValues.Count > 0)
+                            {
+                                field.SelectedValue = field.PossibleValues[0];
+                            }
+                        }
+                    }
+                    #endregion
+                }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs]: Updated

File: Ginger/Ginger/ALM/Repository/RQMRepository.cs [hunk 3]
```diff
@@ -240,7 +241,35 @@ public override bool ExportBusinessFlowToALM(BusinessFlow businessFlow, bool per
             {
                 return false;
             }
-            if (WorkSpace.Instance.Solution.ExternalItemsFields.Where(x => x.ItemType == "TestCase").ToList().Count == 0)
+            ObservableList<ExternalItemFieldBase> OriginalExternalFieldBase = new ObservableList<ExternalItemFieldBase>();
+            if(WorkSpace.Instance.Solution.ALMConfigs.Where(x=>x.DefaultAlm).FirstOrDefault().ALMProjectGUID != WorkSpace.Instance.Solution.ExternalItemsFields.FirstOrDefault().ProjectGuid)
+            {
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/ALM/Repository/RQMRepository.cs:246]: Updated

File: Ginger/Ginger/ALM/Repository/RQMRepository.cs [hunk 4]
```diff
@@ -240,7 +241,35 @@
...
+                    ExternalItemFieldBase externalItemFieldBasetest = new ExternalItemFieldBase();
+                    externalItemFieldBasetest.Name = externalItemFieldBase.Name;
+                    externalItemFieldBasetest.ID = externalItemFieldBase.ID;
+                    externalItemFieldBasetest.ItemType = externalItemFieldBase.ItemType;
+                    externalItemFieldBasetest.Guid = externalItemFieldBase.Guid;
+                    externalItemFieldBasetest.IsCustomField = externalItemFieldBase.IsCustomField;
+                    if(!string.IsNullOrEmpty(WorkSpace.Instance.Solution.ExternalItemsFields.FirstOrDefault(x=>x.Name.Equals(externalItemFieldBase.Name,StringComparison.CurrentCultureIgnoreCase)).SelectedValue))
+                    {
+                        externalItemFieldBasetest.SelectedValue = WorkSpace.Instance.Solution.ExternalItemsFields.FirstOrDefault(x => x.Name.Equals(externalItemFieldBase.Name, StringComparison.CurrentCultureIgnoreCase)).SelectedValue;
+                    }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/ALM/Repository/RQMRepository.cs:259]: done

File: Ginger/Ginger/ALM/Repository/RQMRepository.cs [hunk 5]
```diff
@@ -240,7 +241,35 @@
...
+                    else
+                    {
+                        externalItemFieldBasetest.SelectedValue = externalItemFieldBase.SelectedValue;
+                    }
+                    OriginalExternalFieldBase.Add(externalItemFieldBasetest);
+                }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/ALM/Repository/RQMRepository.cs:265]: done

File: Ginger/GingerCoreNET/ALMLib/RQM/ExportToRQM.cs [hunk 6]
```diff
@@ -905,6 +906,23 @@
...
+                    ACL_Data_Contract.ExternalItemFieldBase field = new ACL_Data_Contract.ExternalItemFieldBase();
+                    field.ItemType = fields[indx].ItemType;
+                    field.Name = fields[indx].Name;
+                    field.SelectedValue = fields[indx].SelectedValue;
+                    fieldsToReturn.Add(field);
+            }
+
+            return fieldsToReturn;
+        }
+
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ExportToRQM.cs:925]: Done

File: Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs [hunk 7]
```diff
@@ -1227,11 +1242,265 @@ public static ObservableList<ExternalItemFieldBase> GetOnlineItemFields(Backgrou
...
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs:1501]: Done

File: Ginger/GingerCoreNET/ALMLib/RQM/ExportToRQM.cs [hunk 8]
```diff
@@ -905,6 +906,23 @@ public bool ExportBusinessFlowToRQM(BusinessFlow businessFlow, ObservableList<Ex
             return false;
         }
 
+        private List<ACL_Data_Contract.ExternalItemFieldBase> CovertExternalFieldsTOACLDataContractfields(ObservableList<ExternalItemFieldBase> fields)
+        {
+            List<ACL_Data_Contract.ExternalItemFieldBase> fieldsToReturn = [];
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ExportToRQM.cs:911]: No need

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**